backend/ws_server/ws_server.py

```diff
@@ -59,73 +65,79 @@ def start_server(address, port):
-
-        if len(connected) == 1:
-            self.sendMessage(utf("you're the first one here! welcome!"), False)
-        elif len(connected) > 1:
-            connected[0].sendMessage(
-                utf("hey, first guy! someone else connected!"), False)
+        if gameloop_client:
+            gameloop_client.sendMessage(format_msg(
+                'a user has connected', MSG.info), False)
+        info("connections: " + str(len(connected)), INFO_ID)
+        self.sendMessage(format_msg(
+            'you have joined the server.', MSG.info), False)
 
     def onMessage(self, payload, isBinary):
-        # for now, only send text messages, no binary yet
         assert isBinary is False
-        if isBinary:
-            # keep this check in for future reference
-            pass
+        as_string = payload.decode('utf8')
+        message = obj_from_json(as_string)
+        m_type = message['type']
+        info('received message (type {}): {}'.format(m_type, as_string), INFO_ID)
+        if m_type == MSG.chat.name:
+            self.handleChat(message)
+        elif m_type == MSG.tower_request.name:
+            self.handleTowerRequest(message)
+        elif m_type == MSG.tower_update.name:
+            self.handleTowerUpdate(message)
+        elif m_type == MSG.game_update.name:
+            self.handleTowerUpdate(message)
+        elif m_type == MSG.identifier.name:
+            self.handleIdentifier(message)
         else:
-            pass
+            info('warning! server does not handle message with type {}'.format(
+                m_type), INFO_ID)
 
-        as_string = payload.decode('utf8')
-        if as_string == GAMELOOP_CLIENT_IDENTIFIER:
-            global gameloop_client
-            gameloop_client = self
-            print("game engine client regisetered!!")
+    def handleIdentifier(self, json_msg):
+        """Handle identification messages, such as the server
+        identification message."""
+        global gameloop_client
+        gameloop_client = self
+        info("game engine client registered", INFO_ID)
 
-        message = obj_from_json(as_string)
-        if message:
-            assert "type" in message
-
-            # TODO, make this not janky
-            if message["type"] == "chat":
-                self.broadcast_message(payload)
-            elif message["type"] == "gameUpdate":
-                self.broadcast_message(payload)
-            elif message["type"] == "towerRequest":
-                global gameloop_client
-                gameloop_client.sendMessage(utf(as_string), False)
-            elif message["type"] == "towerUpdate":
-                self.broadcast_message(payload)
+    def handleChat(self, json_msg):
+        self.broadcast_message(message)
+
+    def handleGameUpdate(self, json_msg):
+        self.broadcast_message(message)
+
+    def handleTowerRequest(self, json_msg):
+        gameloop_client.sendMessage(utf(as_string), False)
```

Comment:
When I switch this around, I get a assorted assert errors (it wants it to be bytes?)

I can get around them by doing something like
> str.encode(str(msg)

But I don't think that's the intent